### ISSUE TITLE
explicit_auto_deref gives a wrong advise

### ISSUE THREAD

**Issue by bachue** [rust-lang/rust-clippy#9123]
### Summary

I have this code

```rust
use std::{ops::Deref, io::Result};

struct A {}

struct AWrapper(A);

impl Deref for AWrapper {
    type Target = A;
    
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

fn returns_a_wrapper() -> Result<AWrapper> {
    todo!()
}

fn accepts_a(_a: &A) {
    todo!()
}

fn main() -> Result<()> {
    accepts_a(&*returns_a_wrapper()?);
    Ok(())
}
```

This code can be compiled now, but `cargo clippy` gives a warning:

```
24 |     accepts_a(&*returns_a_wrapper()?);
   |                ^^^^^^^^^^^^^^^^^^^^^ help: try this: `returns_a_wrapper()?`
   |
   = note: `#[warn(clippy::explicit_auto_deref)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#explicit_auto_deref
```

But if I replace `&*returns_a_wrapper()` by `&returns_a_wrapper()`，then code is failed to be compiled, the error message is 

```
error[[E0308]](https://doc.rust-lang.org/stable/error-index.html#E0308): `?` operator has incompatible types
  --> src/main.rs:24:16
   |
24 |     accepts_a(&returns_a_wrapper()?);
   |                ^^^^^^^^^^^^^^^^^^^^ expected struct `A`, found struct `AWrapper`
   |
   = note: `?` operator cannot convert from `AWrapper` to `A`
```

### Reproducer

I tried this code:

```rust
use std::{ops::Deref, io::Result};

struct A {}

struct AWrapper(A);

impl Deref for AWrapper {
    type Target = A;
    
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

fn returns_a_wrapper() -> Result<AWrapper> {
    todo!()
}

fn accepts_a(_a: &A) {
    todo!()
}

fn main() -> Result<()> {
    accepts_a(&*returns_a_wrapper()?);
    Ok(())
}
```

I expected this code has no warning, but `cargo clippy` gives wrong advise.

### Version

```text
rustc 1.62.0 (a8314ef7d 2022-06-27)
binary: rustc
commit-hash: a8314ef7d0ec7b75c336af2c9857bfaf43002bfc
commit-date: 2022-06-27
host: x86_64-unknown-linux-gnu
release: 1.62.0
LLVM version: 14.0.5
```

**Comment by Jarcho** [comment]
Related rust issue rust_lang/rust#23014
